src/control/ros/RosBarrettHandPositionCommandExecutor.cpp

```diff
@@ -0,0 +1,181 @@
+  goal.command = toJointState(goalPositions,jointNames);
+
+  goal.command.header.stamp = startTime;
+
+  if (!waitForServer())
+    throw std::runtime_error("Unable to connect to action server.");
+
+  {
+    std::lock_guard<std::mutex> lock(mMutex);
+    DART_UNUSED(lock); // Suppress unused variable warning
+
+    if (mInProgress)
+      throw std::runtime_error("Another trajectory is in progress.");
+
+    mPromise = std::promise<void>();
+    mInProgress = true;
+    mGoalHandle = mClient.sendGoal(goal, 
+      boost::bind(&RosBarrettHandPositionCommandExecutor::transitionCallback, this, _1));
+
+    return mPromise.get_future();
+  }
+
+}
+
+//=============================================================================
+// TODO : This is directly reused from RosTrajectoryExecutor.cpp
+// Anyway to reuse this? (assuming we want this here)
+bool RosBarrettHandPositionCommandExecutor::waitForServer()
+{
+  using Clock = std::chrono::steady_clock;
+
+  const auto startTime = Clock::now();
+  const auto endTime = startTime + mConnectionTimeout;
+  auto currentTime = startTime + mConnectionPollingPeriod;
+
+  while (currentTime < endTime)
+  {
+    mCallbackQueue.callAvailable();
+
+    // TODO: Is this thread safe?
+    if (mClient.isServerConnected())
+      return true;
+
+    currentTime += mConnectionPollingPeriod;
+    std::this_thread::sleep_until(currentTime);
+  }
+
+  return false;
+}
+
+//=============================================================================
+void RosBarrettHandPositionCommandExecutor::transitionCallback(GoalHandle handle)
+{
+
+  using actionlib::TerminalState;
+  using Result = pr_control_msgs::SetPositionResult;
+
+  if (handle.getCommState() == actionlib::CommState::DONE)
+  {
+    std::stringstream message;
+    bool isSuccessful = true;
+
+    const auto terminalState = handle.getTerminalState();
+    if (terminalState != TerminalState::SUCCEEDED)
+    {
+      message << "Action " << terminalState.toString();
+
+      const auto terminalMessage = terminalState.getText();
+      if (!terminalMessage.empty())
+        message << " (" << terminalMessage << ")";
+
+      mPromise.set_exception(std::make_exception_ptr(
+          std::runtime_error(message.str())));
+
+      isSuccessful = false;
+    }
+    else{
+      message << "Execution failed.";
+    }
+
+    const auto result = handle.getResult();
+    if (result && result->success == false)
+    {
+      if (!result->message.empty())
+        message << " (" << result->message << ")";
+
+      mPromise.set_exception(std::make_exception_ptr(
+          std::runtime_error(message.str())));
+
+      isSuccessful = false;
+    }
+
+    if (isSuccessful)
+      mPromise.set_value();
+
+    mInProgress = false;
+
+  }
+}
+
+//=============================================================================
+// TODO : This is also basically copied from RosTrajectoryExecutor
+void RosBarrettHandPositionCommandExecutor::step()
```

Comment:
I think that's fine in for now. We may want to write a base class for `actionnlib`-based executors in the future, but there really isn't all that much duplicated code once we move `waitForServer` to a helper function.